src/compiler/checker.ts

```diff
@@ -5217,16 +5228,31 @@ namespace ts {
                 }
                 else if (symbol.valueDeclaration.kind === SyntaxKind.BinaryExpression ||
                          symbol.valueDeclaration.kind === SyntaxKind.PropertyAccessExpression && symbol.valueDeclaration.parent.kind === SyntaxKind.BinaryExpression) {
-                    links.type = getWidenedTypeFromJSSpecialPropertyDeclarations(symbol);
+                    links.type = getWidenedTypeFromJSSpecialPropertyDeclarations(symbol, symbol);
                 }
                 else {
-                    const type = createObjectType(ObjectFlags.Anonymous, symbol);
-                    if (symbol.flags & SymbolFlags.Class) {
-                        const baseTypeVariable = getBaseTypeVariableOfClass(symbol);
-                        links.type = baseTypeVariable ? getIntersectionType([type, baseTypeVariable]) : type;
+                    if (symbol.flags & SymbolFlags.ValueModule && symbol.valueDeclaration && isSourceFile(symbol.valueDeclaration) && symbol.valueDeclaration.commonJsModuleIndicator) {
+                        if (!pushTypeResolution(symbol, TypeSystemPropertyName.Type)) {
+                            return errorType;

```

Comment:
Why not set `links.type` in this case?
This whole function might be neater if it just called `getTypeOfFuncClassEnumModuleUncached` which used `return x` instead of `links.type = x`. Then we would be sure to be setting the value and not forgetting some control flow path.

Reply:
The original reason is that getTypeOfVariableOrParameterOrProperty does. I think that's because you can call `getTypeOfSymbol` in situations that are speculative, and you may successfully be able to get the type at a later time.

I tried this and nothing broke, even when I changed the equivalent line in `getTypeOfVariableOrParameterOrProperty`. So it looks like my guess is wrong. I think I will merge this for now and try that refactor separately, for both.